src/contracts/atlas/NonceManager.sol

```diff
@@ -0,0 +1,280 @@
+//SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.22;
+
+import { AtlasConstants } from "src/contracts/types/AtlasConstants.sol";
+
+contract NonceManager is AtlasConstants {
+    // address => last used sequential nonce
+    mapping(address => uint256) public userSequentialNonceTrackers;
+    mapping(address => uint256) public dAppSequentialNonceTrackers;
+
+    // address => word index => bitmap
+    mapping(address => mapping(uint248 => uint256)) public userNonSequentialNonceTrackers;
+    mapping(address => mapping(uint248 => uint256)) public dAppNonSequentialNonceTrackers;
+
+    // NOTE: To prevent builder censorship, nonces can be processed in any order so long as they aren't duplicated and
+    // as long as the dApp opts in to it.
+
+    // ---------------------------------------------------- //
+    //                                                      //
+    //                     VALIDATION                       //
+    //       Below functions ensure nonces validity         //
+    //                                                      //
+    // ---------------------------------------------------- //
+
+    /// @notice The _handleUserNonces internal function handles the verification of user nonces for both sequential and
+    /// non-sequential nonce systems.
+    /// @param user The address of the user to verify the nonce for.
+    /// @param nonce The nonce to verify.
+    /// @param sequential A boolean indicating if the nonce mode is sequential (true) or not (false)
+    /// @param isSimulation A boolean indicating if the execution is a simulation.
+    /// @return validNonce A boolean indicating if the nonce is valid.
+    function _handleUserNonces(
```

Comment:
`_handleUserNonces` and `_handleDAppNonces` could be a single function where you pass the storage mapping in, rather than hard coding it in the function i think? would reduce some duplication and code size.